dot/parachain/collator-protocol/validator_side.go

```diff
@@ -0,0 +1,316 @@
+	"golang.org/x/exp/slices"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "collator-protocol"))
+
+const (
+	ACTIVITY_POLL              = 10 * time.Millisecond
+	CHECK_COLLATIONS_POLL      = 50 * time.Millisecond
+	MAX_UNSHARED_DOWNLOAD_TIME = 100 * time.Millisecond
+)
+
+var (
+	ErrUnknownOverseerMessage     = errors.New("unknown overseer message type")
+	ErrNotExpectedOnValidatorSide = errors.New("message is not expected on the validator side of the protocol")
+	ErrCollationNotInView         = errors.New("collation is not in our view")
+	ErrPeerIDNotFoundForCollator  = errors.New("peer id not found for collator")
+)
+
+func (cpvs CollatorProtocolValidatorSide) Run(
+	ctx context.Context, OverseerToSubSystem chan any, SubSystemToOverseer chan any) error {
+	inactivityTicker := time.NewTicker(ACTIVITY_POLL)
+
+	for {
+		select {
+		// TODO: polkadot-rust changes reputation in batches, so we do the same?
+		case msg, ok := <-cpvs.OverseerToSubSystem:
+			if !ok {
+				return nil
+			}
+
+			err := cpvs.processMessage(msg)
+			if err != nil {
+				logger.Errorf("processing overseer message: %w", err)
+			}
+		case <-inactivityTicker.C:
+			// TODO: disconnect inactive peers
+			// https://github.com/paritytech/polkadot/blob/8f05479e4bd61341af69f0721e617f01cbad8bb2/node/network/collator-protocol/src/validator_side/mod.rs#L1301
+
+		case unfetchedCollation := <-cpvs.unfetchedCollation:
+			// TODO: If we can't get the collation from given collator within MAX_UNSHARED_DOWNLOAD_TIME,
+			// we will start another one from the next collator.
+
+			// check if this peer id has advertised this relay parent
+			peerData := cpvs.peerData[unfetchedCollation.PendingCollation.PeerID]
+			if peerData.HasAdvertisedRelayParent(unfetchedCollation.PendingCollation.RelayParent) {
+				// if so request collation from this peer id
+				collation, err := cpvs.requestCollation(unfetchedCollation.PendingCollation.RelayParent,
+					unfetchedCollation.PendingCollation.ParaID, unfetchedCollation.PendingCollation.PeerID)
+				if err != nil {
+					logger.Errorf("fetching collation: %w", err)
+				}
+				cpvs.fetchedCollations = append(cpvs.fetchedCollations, *collation)
+			}
+
+		}
+	}
+}
+
+// requestCollation requests a collation from the network.
+// This function will
+// - check for duplicate requests
+// - check if the requested collation is in our view
+func (cpvs CollatorProtocolValidatorSide) requestCollation(relayParent common.Hash,
+	paraID parachaintypes.ParaID, peerID peer.ID) (*parachaintypes.Collation, error) {
+	if !slices.Contains[[]common.Hash](cpvs.ourView.heads, relayParent) {
+		return nil, ErrCollationNotInView
+	}
+
+	// make collation fetching request
+	collationFetchingRequest := CollationFetchingRequest{
+		RelayParent: relayParent,
+		ParaID:      paraID,
+	}
+
+	collationFetchingResponse := NewCollationFetchingResponse()
+	// TODO: find out the appropriate value of collationFetchingResponseTimeout
```

Comment:
AFAIK 30 seconds is the default value for http request timeout, can use that value?

Reply:
polkadot-sdk uses 1.2 sec here. 

https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/network/protocol/src/request_response/mod.rs#L209

https://github.com/paritytech/polkadot-sdk/blob/0a6dfdf973b8e7b669eda6e2ed202fb3549a20b9/polkadot/node/network/protocol/src/request_response/mod.rs#L106